runtime/master/src/main/java/edu/snu/coral/runtime/master/scheduler/SourceLocationAwareSchedulingPolicy.java

```diff
@@ -0,0 +1,178 @@
+        throw new SchedulingException(e);
+      }
+      return attemptSchedule(scheduledTaskGroup, jobStateManager, sourceLocations);
+    }
+  }
+
+  /**
+   * Try to schedule a TaskGroup with source task.
+   * @param scheduledTaskGroup TaskGroup to schedule
+   * @param jobStateManager {@link JobStateManager}
+   * @param jobStateManager jobStateManager which the TaskGroup belongs to.
+   * @return true if the task group is successfully scheduled, false otherwise.
+   */
+  private synchronized boolean attemptSchedule(final ScheduledTaskGroup scheduledTaskGroup,
+                                               final JobStateManager jobStateManager,
+                                               final Set<String> sourceLocations) {
+    final List<ExecutorRepresenter> candidateExecutors =
+        selectExecutorByContainerTypeAndNodeNames(scheduledTaskGroup.getContainerType(), sourceLocations);
+    if (candidateExecutors.size() == 0) {
+      return false;
+    }
+    final int randomIndex = ThreadLocalRandom.current().nextInt(0, candidateExecutors.size());
+    final ExecutorRepresenter selectedExecutor = candidateExecutors.get(randomIndex);
+
+    jobStateManager.onTaskGroupStateChanged(scheduledTaskGroup.getTaskGroupId(), TaskGroupState.State.EXECUTING);
+    selectedExecutor.onTaskGroupScheduled(scheduledTaskGroup);
+    LOG.info("Scheduling {} (source location: {}) to {} (node name: {})", scheduledTaskGroup.getTaskGroupId(),
+        String.join(", ", sourceLocations), selectedExecutor.getExecutorId(),
+        selectedExecutor.getNodeName());
+    return true;
+  }
+
+  @Override
+  public synchronized void onExecutorAdded(final String executorId) {
+    availableExecutors.add(executorId);
+    roundRobinSchedulingPolicy.onExecutorAdded(executorId);
+  }
+
+  @Override
+  public synchronized Set<String> onExecutorRemoved(final String executorId) {
+    availableExecutors.remove(executorId);
+    return roundRobinSchedulingPolicy.onExecutorRemoved(executorId);
+  }
+
+  @Override
+  public void onTaskGroupExecutionComplete(final String executorId, final String taskGroupId) {
+    roundRobinSchedulingPolicy.onTaskGroupExecutionComplete(executorId, taskGroupId);
+  }
+
+  @Override
+  public void onTaskGroupExecutionFailed(final String executorId, final String taskGroupId) {
+    roundRobinSchedulingPolicy.onTaskGroupExecutionFailed(executorId, taskGroupId);
+  }
+
+  /**
+   * @param containerType type of the desired container type
+   * @param nodeNames set of node names
+   * @return list of executors, which resides in one of {@code nodeNames}, has container type of {@code containerType},
+   *         and has an empty slot for execution
+   */
+  private synchronized List<ExecutorRepresenter> selectExecutorByContainerTypeAndNodeNames(
+      final String containerType, final Set<String> nodeNames) {
+    final Map<String, ExecutorRepresenter> executorIdToExecutorRepresenter
+        = containerManager.getExecutorRepresenterMap();
```

Comment:
Unfortunately, directly calling `containerManager.getExecutorRepresenterMap()` may result in race conditions, as the map can be updated concurrently on new executor events. (e.g., executor added)

`RoundRobinSchedulingPolicy` gets around this with a method called `updateCachedExecutorRepresenterMap()`, which is not the best way, but is a method that works.

I think refactoring `executorRepresenterMap` in `ContainerManager` into a separate Tang-injected singleton class would be nice. However, if that cannot be done easily, then I guess a quick solution would be to reuse the cached `executorRepresenterMap` in `RoundRobinSchedulingPolicy`.

@jooykim we'd love your feedback on this. 😄

Reply:
Or how about modifying 
```java
void onExecutorAdded(String executorId);
```
to
```java
void onExecutorAdded(ExecutorRepresenter executorRepresenter);
```
so that SchedulingPolicy can maintian local list of executors?

cc @jooykim

Reply:
I'm not sure about that, as that means we have two copies of the same thing, which makes it harder to reason about concurrency. For example, when an executor is removed, there will be a point in time when one copy exists and the other doesn't.

Reply:
The 'cachedExecutorRepresetnerMap' approach has the same problem. For example, when an executor is removed, there will be a point in time when the executor is removed but the cached map holds obsolete value.

Reply:
+1 for @johnyangk's opinion. I'll create separate injectable registry that maps executorId and executorRepresenter.

Reply:
Done.